Add option to pause proc timer to erlang:suspend_process/2

jcpetruzza opened the pull request:
We add a `pause_proc_timer` option to `erlang:suspend_process/2`. When provided, if the suspendee is waiting on a `receive` statement, the timer associated to the `after` clause will be paused, and resumed when the process is eventually resumed.

The main motivation for this is for a debugger to be able to pause all user processes without leading to timeouts.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=8670) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=8670) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=8670) it.</sub>

jhogberg commented:
Thanks for the PR! Most of us are away on vacation so we have limited time to look at this, and we'll get back to it once we're back from vacation.

(One thing that strikes me at a glance though is that the commits feel a bit too split up: there's no way to reach the code in the first five commits until the last commit. Feel free to squash, or leave it as-is until we're back :-))

jhogberg commented on erts/emulator/beam/erl_hl_timer.c:
```diff
@@ -2714,6 +2760,70 @@ erts_cancel_proc_timer(Process *c_p)
+
+        CANCEL_TIMER(c_p);
+
+        if (ERTS_PTMR_NONE != erts_atomic_cmpxchg_mb(&c_p->common.timer,
```
Would a release barrier be sufficient here?

jcpetruzza replied:
I agree, changed to relb!

jcpetruzza commented on erts/emulator/beam/erl_process.c:
```diff
@@ -7201,6 +7201,10 @@ resume_process(Process *p, ErtsProcLocks locks)
     add2runq(enqueue, enq_prio, p, state, NULL);
+
+    if (state & ERTS_PSFLG_PAUSED_TIMER) {
+        erts_resume_paused_proc_timer(p);
```
This makes the lock-check unhappy, as this calls `erts_proc_sig_send_rpc_request()` which will try to acquire `proc_msgq` while holding `proc_status`. Acquiring both at the same time before calling `resume_process()` doesn't help as we'll then try to lock `proc_msgq` twice. I guess one option would be to extend `erts_proc_sig_send_rpc_request()` with an argument with the locks held, to check if it needs to acquire `proc_msgq` or not. Still, that will not be enough as there are calls to `erts_resume()` under mutexes that can't be acquired before `proc_msgq` (e.g. persistent-term updates). We could try to reorder the locks, but I guess it may be tricky to untangle.

Any recommendation on how to deal with this?

jcpetruzza replied:
Looking into this again; instead of using `erts_proc_sig_send_rpc_request()`, maybe we could use a system-task for this, that only requires holding `proc_status`, I think.

rickard-green commented on erts/emulator/beam/erl_process.c:
```diff
@@ -9006,11 +9067,11 @@ erts_internal_suspend_process_2(BIF_ALIST_2)
             send_sig = !0;
         else {
-            send_sig = !suspend_process(BIF_P, rp);
+            send_sig = !suspend_process(BIF_P, rp, pause_proc_timer);
```
Pause the timer before or after calling `suspend_process()`. Here you are in the context where you can do it right away.

jcpetruzza commented on erts/emulator/beam/erl_process.c:
```diff
@@ -9051,8 +9119,8 @@ resume_process_1(BIF_ALIST_1)
 {
```
I also dislike that the timer gets paused even after all those interested in pausing it, resumed the process. This should be changed. Do we need to add a new option though? I was thinking of adding a new counter in the process structure, similar to `rcount` but with the count of processes that have suspended the proc_timer. When `resume_process()` we know if the caller paused the proc timer as well, so we know if the new counter needs decreasing. When it reaches 0, we could unpause the timer. Would that work?

rickard-green replied:
An option is better since:
* You don't need to add anything to the process structure. We try to avoid adding things to the process structure when it is not frequently used since all processes will have to allocate memory for it even when not used. Since you can have a very large amount of processes, this overhead can become significant. There might be exceptions for important things or when it is not possible to do something another way. This is a pure debug feature that can be done some other way, though.
* The unpause could potentially be done too early. The resume operation corresponding to the suspend operation with pause could be done later than other suspend/resume operations which potentially would unpause the timer too early.

I think you should have a counter in the `ErtsMonitorSuspend` structure, though, and throw a `badarg` error exception in the `erlang:resume_process()` BIF if the amount of suspend with pause and suspend without pause do not match with the amount of resume with unpause and resume without unpause.

jcpetruzza commented on erts/emulator/beam/erl_process.c:
```diff
@@ -7194,16 +7240,19 @@ resume_process(Process *p, ErtsProcLocks locks)
 
     state = erts_atomic32_read_nob(&p->state);
+    if (state & ERTS_PSFLG_PAUSED_TIMER) {
+        schedule_unpause_proc_timer(p->common.id, locks);
```
@rickard-green The i[nitial version of this PR](https://github.com/erlang/otp/commits/36bef84b16691500a3bea10e4a3ceaa7e19453ea) was using `erts_proc_sig_send_rpc_request()`. The reason I changed it to use sys-tasks was that at this point, we'd get a lock-order violation like this:

```
Lock order violation occurred when locking 'proc_msgq:<0.269.0>[proclock]'(none)!                                                                                                                                                                                                                                           
Currently these locks are locked by the scheduler 1 thread:                                                                                                                                                                                                                                                                 
  'proc_main:<0.269.0>[proclock](beam/erl_process.c:10109)'(none)                                                                                                                                                                                                                                                           
  'proc_status:<0.269.0>[proclock](beam/erl_process.c:9300)'(none) 
```

So the problem remains on how to avoid this. We need to somehow send the rpc-request signal before the status lock is acquired

rickard-green replied:
I missed that this one would have a problem as well.

The problem is that the unpausing is triggered from the low level primitive `resume_process()` which handles all resumes regardless of whether corresponding suspend came with a pause or not. Unpausing due to a resume that do not correspond to a suspend with a pause feels like a bug to me.

The only way to pause timers is via the `erlang:suspend_process()` BIF. If the unpausing is done from corresponding resume in the `erlang:resume_process()` BIF, the problem goes away.

jcpetruzza commented on erts/emulator/beam/erl_process.c:
```diff
@@ -9027,6 +9088,13 @@ erts_internal_suspend_process_2(BIF_ALIST_2)
+        Process *rp = erts_proc_lookup(BIF_ARG_1);
+        erts_proc_lock(rp, ERTS_PROC_LOCK_STATUS);
+
+        schedule_pause_proc_timer(rp, ERTS_PROC_LOCK_STATUS);
```
@rickard-green Notice that, afair, this wasn't the point were we run into issues. If I'm not wrong, cancelling a proc timer can be done from any scheduler, so we didn't need a rpc-request signal for that. I think I changed it to use a sys-task in this version only for symmetry, but the initial issue was creating the proc timer while resuming the suspended process (see my other comment).

jcpetruzza commented:
@rickard-green I've pushed a new version that address all previous issues, and indeed is satisfyingly simpler:

  * Each call to `suspend_process()` with the `pause_proc_timer` option requires an eventual `resume_process()` with the `resume_proc_timer` option.
  * The timer gets resumed whenever the count of  pause_proc_timer requests gets to zero, even if the process remains suspended
  * For this, we use a counter in the ErtsPausedProcTimer structure and a new counter in the suspend monitor
  * We rely again only on rpc signals instead of sys-tasks

jcpetruzza commented:
Rebased, fixed two errors, and added a "stress-test" that would surface them. The issues where:
  1. when resuming, the timer needs to be started first, or it will race with receiving a message and clearing the proc timer;
  2. when suspending, if we fail to send the rpc signal, it's because the process was running and exiting; we mustn't delete the monitor in that case as the suspend signal will still be handled

rickard-green commented:
I finally got to this. Sorry for the long wait! I've rewritten it so that we unconditionally pause the timer on `suspend_process()` as we talked about. I've pushed a commit which you can amend with your last commit by running `git rebase --autosquash -i bcaa4ae648e2d6f20578df1747998247f63be917`.

rickard-green commented:
Something went wrong with the amend of the commit. Changes in `erlang.beam` and `erts_internal.beam` are still there. There should not be any changes in those files.

jcpetruzza commented:
Oops, my bad: I updated them while initially rebasing. Pushed a fixed version
